Add Label support for Images (build), Networks and Volumes on Creation

ehazlett opened the pull request:
This adds label support for build, networks and volumes at creation.

Build: `docker build --label foo=bar ...`

Networks: `docker network create --label foo=bar ...`

Volumes: `docker volume create --label foo=bar ...`

You can verify the labels are set by running the respective `inspect` command on the ressource just created.

Note: adding separate PRs to engine-api and libnetwork for corresponding changes

h/t to @tiborvass for helping!!
Closes: #20356

ehazlett committed:
add label support for build, networks and volumes

build: implement --label

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

network: allow adding labels on create

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

volume: allow adding labels on create

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

add tests for build, network, volume

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

vendor: libnetwork and engine-api bump

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

cpuguy83 commented:
On the volumes side, it doesn't look like we're persisting these labels anywhere.

ehazlett commented:
Hmm they should be in the volume store:

```
dev@dev:~/sandbox# /usr/local/bin/docker volume create --label foo=bar --name test
test
dev@dev:~/sandbox# /usr/local/bin/docker volume ls
DRIVER              VOLUME NAME
local               test
dev@dev:~/sandbox# /usr/local/bin/docker volume inspect test
[
    {
        "Name": "test",
        "Driver": "local",
        "Mountpoint": "/var/lib/docker/volumes/test/_data",
        "Labels": {
            "foo": "bar"
        }
    }
]
```

ehazlett commented:
Ah yes I see.  Ok looking.

tiborvass commented on builder/dockerfile/builder.go:
```diff
@@ -212,6 +213,10 @@ func (b *Builder) build(config *types.ImageBuildOptions, context builder.Context
 	var shortImgID string
 	for i, n := range b.dockerfile.Children {
+		if i == len(b.dockerfile.Children) {
+			b.isLast = true
```
@ehazlett looking at the code it seems suspicious that it's not being used anywhere.

ehazlett replied:
good catch -- i'll clean that

thaJeztah commented:
just a quick check; does this also support labels without an implicit value set, like we support on `docker run`? i.e. `docker volume create --label foo` (which will set label `foo` with value `""` (empty string))

ehazlett commented:
@thaJeztah correct

```
root@dev:~/sandbox# docker volume create --label foo --name temp                                                                                                                                                   
temp
root@dev:~/sandbox# docker volume inspect temp
[
    {
        "Name": "temp",
        "Driver": "local",
        "Mountpoint": "/var/lib/docker/volumes/temp/_data",
        "Labels": {
            "foo": ""
        }
    }
]
```

calavera commented:
@ehazlett how does this work with external plugins? It doesn't look like you're adding the labels to the external volume requests. Am I missing something?

cpuguy83 commented:
@calavera That's correct, this was a conscious decision to ensure consistency in label behavior. We don't send labels to drivers for anything else either, IIRC because labels are meant for higher-level abstractions not lower-level.

I think for now it makes sense to store locally on the daemon since there is no current concept of a multi-host volume even if some drivers implement it.

calavera commented:
> I think for now it makes sense to store locally on the daemon since there is no current concept of a multi-host volume even if some drivers implement it.

SGTM

calavera commented:
Design LGTM

cpuguy83 commented:
Also ok on design here, ping @mavenugo for networks.

vdemeester commented:
@ehazlett Quick question about the `--label` on `build` : we add this flag for consistency with other commands I guess ? 

There is something weird with the cache : given the following `Dockerfile`

```Dockerfile
FROM busybox
LABEL foo=bar
RUN ls
```

Let's build it different way to see how it behave.

```bash
# Build first without label and the with
$ docker build -t test .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox
 ---> 8b17efa7e876
Step 2 : LABEL foo bar
 ---> Running in 5987eeb9ec11
 ---> 85959571113f
Removing intermediate container 5987eeb9ec11
Step 3 : RUN ls
 ---> Running in 107f0cc0146e
bin
dev
etc
home
proc
root
sys
tmp
usr
var
 ---> e16e1200579b
Removing intermediate container 107f0cc0146e
Successfully built e16e1200579b
$ docker inspect -f "{{json .Config.Labels}}" test                                                                                             
{"foo":"bar"}
$ docker build --label foo=baz -t test .                                                                                                       
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox
 ---> 8b17efa7e876
Step 2 : LABEL foo bar
 ---> Using cache
 ---> 85959571113f
Step 3 : RUN ls
 ---> Using cache
 ---> e16e1200579b
$ docker inspect -f "{{json .Config.Labels}}" test
{"foo":"bar"}

# Clean the image
$ docker rmi test

# Build first with label and the without
$ docker build --label foo=baz -t test .                                                                                                       
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox
 ---> 8b17efa7e876
Step 2 : LABEL foo bar
 ---> Running in 0413afce32a3
 ---> 4ed22a0eb99d
Removing intermediate container 0413afce32a3
Step 3 : RUN ls
 ---> Running in 4bc6e81b1cf0
bin
dev
etc
home
proc
root
sys
tmp
usr
var
 ---> c7b3edb1e6c7
Removing intermediate container 4bc6e81b1cf0
Successfully built c7b3edb1e6c7
$ docker inspect -f "{{json .Config.Labels}}" test                                                                                             
{"foo":"baz"}
$ docker build -t test .                                                                                                                       
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox
 ---> 8b17efa7e876
Step 2 : LABEL foo bar
 ---> Running in 83dba90f5411
 ---> fff9ee594434
Removing intermediate container 83dba90f5411
Step 3 : RUN ls
 ---> Running in 8ad59a2b64f2
bin
dev
etc
home
proc
root
sys
tmp
usr
var
 ---> 398900d4e3fa
Removing intermediate container 8ad59a2b64f2
Successfully built 398900d4e3fa
$ docker inspect -f "{{json .Config.Labels}}" test
{"foo":"bar"}
$ docker build --label foo=biz -t test .                                                                                                       
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox
 ---> 8b17efa7e876
Step 2 : LABEL foo bar
 ---> Using cache
 ---> fff9ee594434
Step 3 : RUN ls
 ---> Using cache
 ---> 398900d4e3fa
Successfully built 398900d4e3fa
$ docker inspect -f "{{json .Config.Labels}}" test
{"foo":"bar"}

# Clean images
$ docker rmi test

# Another try
$ docker build --label foo=baz -t test .                                                                                                       
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox
 ---> 8b17efa7e876
Step 2 : LABEL foo bar
 ---> Running in b7ae803c1780
 ---> 8c67745bbf7f
Removing intermediate container b7ae803c1780
Step 3 : RUN ls
 ---> Running in 5eb42ef8b39a
bin
dev
etc
home
proc
root
sys
tmp
usr
var
 ---> f8abbbda2203
Removing intermediate container 5eb42ef8b39a
Successfully built f8abbbda2203
$ docker inspect -f "{{json .Config.Labels}}" test                                                                                             
{"foo":"baz"}
$ docker build --label foo=biz -t test .                                                                                                       
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox
 ---> 8b17efa7e876
Step 2 : LABEL foo bar
 ---> Running in 7db61c4c1ce0
 ---> a3b417a61254
Removing intermediate container 7db61c4c1ce0
Step 3 : RUN ls
 ---> Running in 0bf5a91425af
bin
dev
etc
home
proc
root
sys
tmp
usr
var
 ---> 708e70ccb416
Removing intermediate container 0bf5a91425af
Successfully built 708e70ccb416
$ docker inspect -f "{{json .Config.Labels}}" test
{"foo":"biz"}
```

If the image has been build at least once without the `--label` then it uses the cache and `--label` has no effect, it does not override the label. I'm not sure if it's a behavior we want or not :angel:.

ehazlett commented:
@vdemeester great point.  I think we would probably want to update the labels if specified.  Thoughts @tiborvass @calavera

thaJeztah commented:
@ehazlett tests are failing;

```
18:20:31 # github.com/docker/docker/daemon
18:20:31 daemon/daemon_test.go:124: not enough arguments in call to "github.com/docker/docker/volume/store".New
18:20:31 daemon/daemon_test.go:124: multiple-value "github.com/docker/docker/volume/store".New() in single-value context

...

18:21:44 # github.com/docker/docker/volume/store
18:21:44 volume/store/store_test.go:15: not enough arguments in call to New
18:21:44 volume/store/store_test.go:15: multiple-value New() in single-value context
18:21:44 volume/store/store_test.go:43: not enough arguments in call to New
18:21:44 volume/store/store_test.go:43: multiple-value New() in single-value context
18:21:44 volume/store/store_test.go:74: not enough arguments in call to New
18:21:44 volume/store/store_test.go:74: multiple-value New() in single-value context
18:21:44 volume/store/store_test.go:91: not enough arguments in call to New
18:21:44 volume/store/store_test.go:91: multiple-value New() in single-value context
18:21:44 volume/store/store_test.go:106: not enough arguments in call to New
18:21:44 volume/store/store_test.go:106: multiple-value New() in single-value context
18:21:44 volume/store/store_test.go:106: too many errors
```

ehazlett commented:
@thaJeztah thx on it

ehazlett commented:
Ok I think we're getting close.  I tried to update vendor but received this:

```
vendor/src/github.com/docker/notary/trustmanager/x509utils.go:21:2: no buildable Go source files in /go/src/github.com/docker/docker/vendor/src/github.com/agl/ed25519
vendor/src/github.com/docker/notary/tuf/data/keys.go:17:2: cannot find package "github.com/docker/go/canonical/json" in any of:
    /go/src/github.com/docker/docker/vendor/github.com/docker/go/canonical/json (vendor tree)
    /usr/local/go/src/github.com/docker/go/canonical/json (from $GOROOT)
    /go/src/github.com/docker/go/canonical/json (from $GOPATH)
    /go/src/github.com/docker/docker/vendor/src/github.com/docker/go/canonical/json
```

The win2lin test is failing with a disk space error.

I will take a look to see if I can track down the vendoring issue.

mavenugo commented:
@ehazlett libnetwork tag v0.7.0-dev.9 is available with the network labels support that you added. Please use it to vendor in the libnetwork changes.

ehazlett commented:
@mavenugo after doing this i get the follow diff.  Upon build/test it gives me the error above since it is removing `docker/go`

```
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   hack/vendor.sh
	deleted:    vendor/src/github.com/agl/ed25519/ed25519.go
	deleted:    vendor/src/github.com/agl/ed25519/edwards25519/const.go
	deleted:    vendor/src/github.com/agl/ed25519/edwards25519/edwards25519.go
	modified:   vendor/src/github.com/docker/engine-api/client/image_build.go
	modified:   vendor/src/github.com/docker/engine-api/types/client.go
	modified:   vendor/src/github.com/docker/engine-api/types/types.go
	deleted:    vendor/src/github.com/docker/go/canonical/json/decode.go
	deleted:    vendor/src/github.com/docker/go/canonical/json/encode.go
	deleted:    vendor/src/github.com/docker/go/canonical/json/fold.go
	deleted:    vendor/src/github.com/docker/go/canonical/json/indent.go
	deleted:    vendor/src/github.com/docker/go/canonical/json/scanner.go
	deleted:    vendor/src/github.com/docker/go/canonical/json/stream.go
	deleted:    vendor/src/github.com/docker/go/canonical/json/tags.go
	modified:   vendor/src/github.com/docker/libnetwork/CHANGELOG.md
	modified:   vendor/src/github.com/docker/libnetwork/Dockerfile.build
	modified:   vendor/src/github.com/docker/libnetwork/network.go
	modified:   vendor/src/github.com/docker/libnetwork/osl/interface_linux.go
	deleted:    vendor/src/github.com/miekg/pkcs11/.gitignore
	deleted:    vendor/src/github.com/miekg/pkcs11/.travis.yml
	deleted:    vendor/src/github.com/miekg/pkcs11/README.md
	deleted:    vendor/src/github.com/miekg/pkcs11/const.go
	deleted:    vendor/src/github.com/miekg/pkcs11/error.go
	deleted:    vendor/src/github.com/miekg/pkcs11/hsm.db
	deleted:    vendor/src/github.com/miekg/pkcs11/pkcs11.go
	deleted:    vendor/src/github.com/miekg/pkcs11/pkcs11.h
	deleted:    vendor/src/github.com/miekg/pkcs11/pkcs11f.h
	deleted:    vendor/src/github.com/miekg/pkcs11/pkcs11t.h
	deleted:    vendor/src/github.com/miekg/pkcs11/softhsm.conf
	deleted:    vendor/src/github.com/miekg/pkcs11/types.go
```

```
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
vendor/src/github.com/docker/notary/trustmanager/x509utils.go:21:2: no buildable Go source files in /go/src/github.com/docker/docker/vendor/src/github.com/agl/ed25519
vendor/src/github.com/docker/notary/tuf/data/keys.go:17:2: cannot find package "github.com/docker/go/canonical/json" in any of:
	/go/src/github.com/docker/docker/vendor/github.com/docker/go/canonical/json (vendor tree)
	/usr/local/go/src/github.com/docker/go/canonical/json (from $GOROOT)
	/go/src/github.com/docker/go/canonical/json (from $GOPATH)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/go/canonical/json
```

mavenugo commented:
@ehazlett we are seeing a few issues with the hack/vendor.sh script that manifests in picking up / removing unrelated deps. Whenever I vendor in particular dep, I take care of this manually using `git checkout` on the unrelated changes. In this case, the libnetwork change triggered vendor.sh update should include just the libnetwork changes and not other deps (pls `git checkout` them).

ehazlett commented:
Ok I will fixup manually. Thanks.
On Mar 20, 2016 16:17, "Madhu Venugopal" <notifications@github.com> wrote:

> @ehazlett <https://github.com/ehazlett> we are seeing a few issues with
> the hack/vendor.sh script that manifests in picking up / removing unrelated
> deps. Whenever I vendor in particular dep, I take care of this manually
> using git checkout on the unrelated changes. In this case, the libnetwork
> change triggered vendor.sh update should include just the libnetwork
> changes and not other deps (pls git checkout them).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/21270#issuecomment-199009493>
>

thaJeztah commented:
oh boy, @ehazlett looks like this needs a rebase now :cry:

mavenugo commented:
@ehazlett the vendor-in commit is not picking up all the libnetwork specific changes. But the changes are spread across other commits. It would be great if you can organize the changes in the way that the vendor-in commit brings in the changes from the dependent projects.

ehazlett commented:
I just squashed the commits to put everything together.

vdemeester commented:
@ehazlett arf, I'm gonna be a pain in the ass but, we usually split commit that do the vendor from the other ones ; so it would be perfect if you had a commit with the vendoring changes, and one with the actual code of the PR :stuck_out_tongue_closed_eyes:

tiborvass commented:
@mavenugo @vdemeester sorry guys we don't have time to please everyone :) I suggest someone opens a PR to the vendoring policy (`VENDORING.md`) and clarifies the situation. It seems to me all solutions have their problems.

I asked @ehazlett to just squash it all into one commit and move on. Hopefully we can get more testing of this PR and less bikeshedding :)

vdemeester commented:
@tiborvass ok, sounds good to me then :sweat_smile:

tiborvass commented:
LGTM

icecrime commented:
Please give us your feedback!

- [ ] @mavenugo: ok for networking?
- [ ] @calavera @cpuguy83: ok for volumes?
- [ ] @vdemeester: overall LGTM? :innocent: 

Thanks

vdemeester commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -6596,6 +6596,127 @@ func (s *DockerSuite) TestBuildWorkdirWindowsPath(c *check.C) {
+	}
+}
+
+func (s *DockerSuite) TestBuildLabelCacheCommit(c *check.C) {
```
What is the difference between this test and the one above ?
If the purpose of this test is to *exercise* the cache, shouldn't we build the image twice ?

ehazlett replied:
The first one says to use the build cache and the second ignores.  This is to make sure the label is applied during the commit.

vdemeester replied:
@ehazlett ah ok, I missed the `useCache` boolean argument ; the first one forces docker to not use the cache, and the seconds let's docker use it if required. My question is then : does this test `TestBuildLabelCacheCommit` make the assumption that something is already in the cache ? 
- If true, this means that we need to *fill the cache* first as after each test `DockerSuite.TearDownTest` is ran and it deletes all non-protected images, meaning that the cache is pretty much empty :sweat_smile:.
- If not, I'm not sure I understand the point of the test (as it's basically the same as ˙TestBuildLabel` — maybe I'm missing something though.

ehazlett replied:
Ah yes it needs to be filled.  I'll look.  Thx!

ehazlett replied:
@vdemeester ok i think i have it filled now.

vdemeester replied:
@ehazlett yes !

vdemeester commented:
A quick question on one of the integration test. Other than that I think it looks good :+1:.
So you have my *overall* LGTM 🐲

mavenugo commented:
Tested the changes with overlay driver and I confirm that the labels are distributed across the cluster and the `docker network inspect` in all the nodes reflect the label.

LGTM

vdemeester commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -6596,6 +6596,134 @@ func (s *DockerSuite) TestBuildWorkdirWindowsPath(c *check.C) {
+	if _, err := buildImage(name, `
+  FROM `+minimalBaseImage()+`
+  LABEL default foo
+  `, false, []string{"--label", testLabel}...); err != nil {
```
@ehazlett I'm not sure if it's necessary to add the `--label` flag for this first build (and without it it will even cover the use case I describe abvoe :wink:)

ehazlett commented:
@vdemeester ok i removed the label from that first test.  it should be good now.

calavera commented on api/client/network.go:
```diff
@@ -51,6 +52,7 @@ func (cli *DockerCli) CmdNetworkCreate(args ...string) error {
 	cmd.Var(flIpamAux, []string{"-aux-address"}, "auxiliary ipv4 or ipv6 addresses used by Network driver")
 	cmd.Var(flOpts, []string{"o", "-opt"}, "set driver specific options")
 	cmd.Var(flIpamOpt, []string{"-ipam-opt"}, "set IPAM driver specific options")
+	cmd.Var(&flLabels, []string{"-label"}, "set meta data on a network")
```
`meta data` is just one word `metadata` same everywhere else :grin:

ehazlett replied:
shit i knew that.  ok i'll update PR wide.  thx.

calavera commented on builder/dockerfile/builder.go:
```diff
@@ -215,6 +215,9 @@ func (b *Builder) build(config *types.ImageBuildOptions, context builder.Context
 
 	var shortImgID string
 	for i, n := range b.dockerfile.Children {
+		if i == len(b.dockerfile.Children)-1 {
```
can we have a comment here, so we remember we only want to add labels to the last layer?

calavera commented:
Looks good so far, I only have a couple of minor comments. Be aware that #21370 is vendoring a newer version of engine-api.

ehazlett commented:
@calavera ok i've updated those with metadata.  i will submit separate PRs to engine-api and libnetwork for the updates to those.  i'll catch the new vendoring after that PR then.  thx

ehazlett commented:
FYI https://github.com/docker/engine-api/pull/167 sent for engine-api side.

ehazlett commented:
@calavera ok i think i addressed all of yours (updated vendor for engine-api)

cpuguy83 commented:
LGTM

icecrime commented:
Thanks :green_heart:

vdemeester commented:
There is some retwork-related failure :sweat: 

```
07:18:14 ----------------------------------------------------------------------
07:18:14 FAIL: docker_cli_network_unix_test.go:1145: DockerNetworkSuite.TestDockerNetworkConnectPreferredIP
07:18:14 
07:18:14 docker_cli_network_unix_test.go:1157:
07:18:14     verifyIPAddresses(c, "c0", "n0", "172.28.99.88", "2001:db8:1234::9988")
07:18:14 docker_cli_network_unix_test.go:1249:
07:18:14     c.Assert(strings.TrimSpace(out), check.Equals, ipv6)
07:18:14 ... obtained string = ""
07:18:14 ... expected string = "2001:db8:1234::9988"
07:18:14 
07:18:15 
07:18:15 ----------------------------------------------------------------------
07:18:15 FAIL: docker_cli_network_unix_test.go:1181: DockerNetworkSuite.TestDockerNetworkConnectPreferredIPStoppedContainer
07:18:15 
07:18:15 docker_cli_network_unix_test.go:1197:
07:18:15     verifyIPAddresses(c, "c0", "n0", "172.30.55.44", "2001:db8:abcd::5544")
07:18:15 docker_cli_network_unix_test.go:1249:
07:18:15     c.Assert(strings.TrimSpace(out), check.Equals, ipv6)
07:18:15 ... obtained string = ""
07:18:15 ... expected string = "2001:db8:abcd::5544"
07:18:15 
07:18:16 
07:18:16 ----------------------------------------------------------------------
```

https://jenkins.dockerproject.org/job/Docker-PRs/25585/console

ping @mavenugo

mavenugo commented:
@vdemeester @ehazlett the failure is due to the recent changes in `dev.10` to the ipv6 handling.

@aboch can you PTAL asap ?

@ehazlett we don't want to block you with this one. Can you please revert the libnetwork vendor-in to `dev.9` tag and proceed. We will get this resolved (either in the IT or the libnetwork code) and re-vendor libnetwork.

vdemeester commented:
Thanks for looking into that @mavenugo :wink:

aboch commented:
@ehazlett 
If you still want to vendor libnetwork v0.7.0-dev.10, the fix for the failing tests is to add the `"--ipv6"` argument in the dockerCmd() for `docker network create`.

ehazlett commented:
@aboch sorry just reset to v0.7.0-dev.9 and rebased.

tiborvass commented:
@mavenugo still the same :(

ehazlett commented:
crap i think this is my fault  -- fixing

mavenugo commented:
@ehazlett thanks.
@tiborvass we vendored https://github.com/docker/docker/pull/21396 independently.
